alf/environments/wrappers.py

```diff
@@ -102,3 +104,200 @@ def step(self, action):
+        super().__init__(env)
+        if frame_skip <= 0:
+            raise ValueError(
+                'Frame skip should be strictly positive, got {}'.format(frame_skip))
+        if screen_size <= 0:
+            raise ValueError('Target screen size should be strictly positive, got {}'
+                            .format(screen_size))
+
+        self.terminal_on_life_loss = terminal_on_life_loss
+        self.frame_skip = frame_skip
+        self.screen_size = screen_size
+        self.noop_max = noop_max
+        assert env.unwrapped.get_action_meanings()[0] == "NOOP"
+
+        obs_dims = self.env.observation_space
+        # Stores temporary observations used for pooling over two successive
+        # frames.
+        self.screen_buffer = [
+            np.empty((obs_dims.shape[0], obs_dims.shape[1]), dtype=np.uint8),
+            np.empty((obs_dims.shape[0], obs_dims.shape[1]), dtype=np.uint8)
+        ]
+
+        self.observation_space = gym.spaces.Box(
+            low=0,
+            high=255,
+            shape=(self.screen_size, self.screen_size, 1),
+            dtype=np.uint8)
+
+        self.game_over = True
+        self.lives = 0  # Will need to be set by reset().
+
+    def reset_with_random_noops(self):
+        self.env.reset()
+        n_noops = self.env.unwrapped.np_random.randint(
+            1, self.noop_max + 1)
+        for _ in range(n_noops):
+            _, _, game_over, _ = self.env.step(0)
+            if game_over:
+                self.env.reset()
+
+    def fire_on_reset(self):
+        # The following code is from https://github.com/openai/gym/...
+        # ...blob/master/gym/wrappers/atari_preprocessing.py
+        action_meanings = self.env.unwrapped.get_action_meanings()
+        if action_meanings[1] == 'FIRE' and len(action_meanings) >= 3:
+            self.env.step(1)
+            self.env.step(2)
+
+    def reset(self):
+        """Resets the environment.
+
+        Returns:
+        observation: numpy array, the initial observation emitted by the
+            environment.
+        """
+        if self.game_over:
+            self.reset_with_random_noops()
+            self.lives = self.env.ale.lives()
+            self.game_over = False
+        else:
+            # reset called because of losing a life
+            self.env.step(0) # no-op
+
+        self.fire_on_reset()
+
+        # in either case, we need to clear the screen buffer
+        self._fetch_grayscale_observation(self.screen_buffer[0])
+        self.screen_buffer[1].fill(0)
+        return self._pool_and_resize()
+
+    def step(self, action):
+        """Applies the given action in the environment.
+
+        Remarks:
+
+        * If a terminal state (from life loss or episode end) is reached, this may
+            execute fewer than self.frame_skip steps in the environment.
+        * Furthermore, in this case the returned observation may not contain valid
+            image data and should be ignored.
+
+        Args:
```

Comment:
comment should follow the format of other comment in alf.

Reply:
updated